tools/ideabuck/src/com/facebook/buck/intellij/ideabuck/run/BuckBuildBeforeRun.java

```diff
@@ -0,0 +1,182 @@
+  static private String tempFilePath;
+
+  public BuckBuildBeforeRun(@NotNull Project project) {
+    super(SYSTEM_ID, project, ID);
+  }
+
+  public void setTempFileOutput(String tempFileOutput) {
+    tempFilePath = tempFileOutput;
+  }
+
+  @Override
+  public String getName() {
+    return "Buck Build";
+  }
+
+  @Override
+  public String getDescription(ExternalSystemBeforeRunTask task) {
+    return "Buck Build";
+  }
+
+  @NotNull
+  @Override
+  public ExternalSystemBeforeRunTask createTask(@NotNull RunConfiguration runConfiguration) {
+    return new ExternalSystemBeforeRunTask(ID, SYSTEM_ID);
+  }
+
+  @Nullable
+  @Override
+  public Icon getIcon() {
+    return BuckIcons.BUCK_TOOL_WINDOW_ICON;
+  }
+
+  @Override
+  public boolean canExecuteTask(
+      RunConfiguration configuration, ExternalSystemBeforeRunTask beforeRunTask
+  ) {
+    if (!(configuration instanceof ApplicationConfiguration)) {
+      return false;
+    }
+    VirtualFile containingFile = ((ApplicationConfiguration) configuration).getMainClass()
+        .getContainingFile().getVirtualFile();
+    VirtualFile buckFile = BuckFileUtil.getBuckFile(containingFile);
+
+    return buckFile != null;
+  }
+
+  @Override
+  public boolean executeTask(
+      final DataContext context,
+      RunConfiguration configuration,
+      ExecutionEnvironment env,
+      ExternalSystemBeforeRunTask beforeRunTask
+  ) {
+
+    Project currentProject = configuration.getProject();
+
+    AtomicBoolean successfulBuild = new AtomicBoolean(false);
+    if (configuration instanceof ApplicationConfiguration) {
+      VirtualFile containingFile = ((ApplicationConfiguration) configuration).getMainClass()
+          .getContainingFile().getVirtualFile();
+
+      VirtualFile buckFile = BuckFileUtil.getBuckFile(containingFile);
+
+      if (buckFile == null) {
+        return false;
+      }
+
+      BuckBuildManager buildManager = BuckBuildManager.getInstance(currentProject);
+      BuckQueryCommandHandler buckQueryCommandHandler =
+          getBuckQueryCommandHandler(currentProject, buckFile);
+      buckQueryCommandHandler.command().addParameter("owner(" + containingFile.getPath() + ")");
+      buildManager.runInCurrentThreadPostEnd(
+          buckQueryCommandHandler,
+          () -> {
+            for (String target : targetResults) {
+              buildTarget(currentProject, successfulBuild, buckFile, buildManager, target);
+            }
+          });
+    }
+    return successfulBuild.get();
+  }
+
+  private void buildTarget(Project currentProject, AtomicBoolean successfulBuild,
+      VirtualFile buckFile, BuckBuildManager buildManager, String target) {
+    BuckBuildCommandHandler buckBuildCommandHandler =
+        new BuckBuildCommandHandler(
+            currentProject,
+            buckFile.getParent(),
+            BuckCommand.BUILD,
+            true
+        );
+    buckBuildCommandHandler.command()
+        .addParameters("--deep", target, "--build-report", tempFilePath);
```

Comment:
Observation of a potential edge case:  if you're using a build server, then building `--deep` may not be sufficient to make your code runnable.  IIUC, asking buck to build `--deep` doesn't require it to build all of your dependencies -- it is permitted to fetch ABIs from the build server, instead, if that is sufficient to build the target you've selected.